File: sdk/typescript/src/client/address.ts
/**
 * External gestaltd address normalization and validation.
 */

export function normalizeAddress(address: string | URL): string {
  const parsed = address instanceof URL ? address : parseAddressString(address);
  if (parsed.protocol !== "http:" && parsed.protocol !== "https:") {
    throw new Error(
      `address must use http or https: ${JSON.stringify(parsed.protocol)}`,
    );
  }
  return parsed.toString().replace(/\/$/, "");
}

function parseAddressString(address: string): URL {
  const trimmed = address.trim();
  if (!trimmed) {
    throw new Error("address is required");
  }
  try {
    return new URL(trimmed);
  } catch {
    throw new Error(`address must be an absolute URL: ${JSON.stringify(address)}`);
  }
}

File: sdk/python/gestalt/public/address.py
"""External gestaltd address normalization and validation."""

from __future__ import annotations

from urllib.parse import urlparse


def normalize_address(address: str) -> str:
    trimmed = address.strip()
    if not trimmed:
        raise ValueError("address is required")
    parsed = urlparse(trimmed)
    if not parsed.scheme or not parsed.netloc:
        raise ValueError(f"address must be an absolute URL: {address!r}")
    if parsed.scheme not in {"http", "https"}:
        raise ValueError(f"address must use http or https: {parsed.scheme!r}")
    return trimmed.rstrip("/")
